rr_platform/src/bigoli_motor_relay_node/bigoli_ethernet_drive_relay.cpp

```diff
@@ -0,0 +1,170 @@
+#include <ros/ros.h>
+#include <rr_platform/speed.h>
+#include <rr_platform/steering.h>
+#include <rr_platform/chassis_state.h>
+#include <boost/asio.hpp>
+#include <boost/array.hpp>
+#include <iostream>
+
+/*@NOTE THIS CODE USES BOOST 1.58 as it is the version currently installed with ROS.
+  If that changes, this code will need to be updated as such!
+  So don't fear if it breaks, just fix the things by checking the links below to examples given.
+*/
+
+using namespace std;
+using boost::asio::ip::tcp;
+
+struct PIDConst {
+  float p;
+  float i;
+  float d;
+};
+
+struct PIDConst accelDrivePID, decelDrivePID, steeringPID;
+
+double speed = 0.0;
+double steeringAngle = 0.0;
+
+double maxAngleMsg;
+const double maxOutput = 1.0;  //#TODO: magic # to launch param maybe???
+
+std::shared_ptr<tcp::socket> currentSocket;
+
+void speedCallback(const rr_platform::speed::ConstPtr &msg) {
+    speed = msg->speed;
+}
+
+void steerCallback(const rr_platform::steering::ConstPtr &msg) {
+    steeringAngle = msg->angle / maxAngleMsg * maxOutput; // Taken from old relay
+}
+
+string readMessage() {
+  //read data from TCP connection
+  boost::array<char, 128> buf;
+  boost::system::error_code error;
+
+  size_t len = currentSocket->read_some(boost::asio::buffer(buf), error);
+  string reading(buf.begin(), buf.end()); //convert buffer into useable string
+
+  if (error == boost::asio::error::eof) {
+    // Connection closed cleanly by peer
+    ROS_ERROR_STREAM("TCP Connection closed by peer: Disconnecting");
+  } else if (error) {
+    ROS_ERROR_STREAM("TCP ERROR: Disconnecting");
+    throw boost::system::system_error(error); // Some other error
+  }
+
+  return reading;
+}
+
+void sendMessage(string message) {
+  boost::array<char, 128> buf;
+  boost::system::error_code error;
+
+  //write data to TCP connection
+  boost::asio::write(*currentSocket, boost::asio::buffer(message), error); //#TODO: error check????
+}
+
+string buildPIDMessage(struct PIDConst pid) {
```

Comment:
What does it mean to pass the keyword "struct" here? Also, it does not affect the performance in this case, but structs should be passed by const reference, like void func(const BigType& data)